letsencrypt/acme/messages2.py

```diff
@@ -0,0 +1,276 @@
+
+
+class _Constant(jose.JSONDeSerializable):
+    """ACME constant."""
+    __slots__ = ('name',)
+    POSSIBLE_NAMES = NotImplemented
+
+    def __init__(self, name):
+        self.POSSIBLE_NAMES[name] = self
+        self.name = name
+
+    def to_json(self):
+        return self.name
+
+    @classmethod
+    def from_json(cls, value):
+        if value not in cls.POSSIBLE_NAMES:
+            raise jose.DeserializationError(
+                '{0} not recognized'.format(cls.__name__))
+        return cls.POSSIBLE_NAMES[value]
+
+    def __repr__(self):
+        return '{0}({1})'.format(self.__class__.__name__, self.name)
+
+    def __eq__(self, other):
+        return isinstance(other, type(self)) and other.name == self.name
+
+
+class Status(_Constant):
+    """ACME "status" field."""
+    POSSIBLE_NAMES = {}
+STATUS_UNKNOWN = Status('unknown')
+STATUS_PENDING = Status('pending')
+STATUS_PROCESSING = Status('processing')
+STATUS_VALID = Status('valid')
+STATUS_INVALID = Status('invalid')
+STATUS_REVOKED = Status('revoked')
+
+
+class IdentifierType(_Constant):
+    """ACME identifier type."""
+    POSSIBLE_NAMES = {}
+IDENTIFIER_FQDN = IdentifierType('dns')  # IdentifierDNS in Boulder
+
+
+class Identifier(jose.JSONObjectWithFields):
+    """ACME identifier."""
+    typ = jose.Field('type', decoder=IdentifierType.from_json)
+    value = jose.Field('value')
+
+
+class Resource(jose.ImmutableMap):
+    """ACME Resource.
+
+    :param body: Resource body.
+    :type body: Instance of `ResourceBody` (subclass).
+
+    :param str uri: Location of the resource.
+
+    """
+    __slots__ = ('body', 'uri')
+
+
+class ResourceBody(jose.JSONObjectWithFields):
+    """ACME Resource body."""
+
+
+class RegistrationResource(Resource):
+    """Registration resource.
+
+    :ivar body: `Registration`
+    :ivar str uri: URI of the resource.
+    :ivar new_authzr_uri: URI found in the 'next' Link header
+
+    """
+    __slots__ = ('body', 'uri', 'new_authzr_uri', 'terms_of_service')
+
+
+class Registration(ResourceBody):
+    """Registration resource body."""
+
+    # on new-reg key server ignores 'key' and populates it based on
+    # JWS.signature.combined.jwk
+    key = jose.Field('key', omitempty=True, decoder=jose.JWK.from_json)
+    contact = jose.Field('contact', omitempty=True, default=())
+    recovery_token = jose.Field('recoveryToken', omitempty=True)
+    agreement = jose.Field('agreement', omitempty=True)
+
+
+class ChallengeResource(Resource, jose.JSONObjectWithFields):
+    """Challenge resource.
+
+    :ivar body: `.challenges.ChallengeBody`
+    :ivar authzr_uri: URI found in the 'up' Link header.
+
+    """
+    __slots__ = ('body', 'authzr_uri')
+
+    @property
+    def uri(self):  # pylint: disable=missing-docstring,no-self-argument
+        # bug? 'method already defined line None'
+        # pylint: disable=function-redefined
+        return self.body.uri
+
+
+class ChallengeBody(ResourceBody):
```

Comment:
I might suggest placing "Body" before Authorization and Registration in order to stay consistent with the rest of the classes within this file.  I think the TODO is probably the right way to go in the future, and I wouldn't mind seeing Annotated challenges having the metadata in the challenge themselves... (This would remove the need for any index caching (IndexedChalls))